File: ObserverCPP/Source.cpp
// Copyright (c) 2022 RodKingroo

#include <iostream>
#include <list>

class iObserver
{
public:
	virtual void Update(float temperature, float himidity, float min_temperature, float max_temperature, float pressure, float feels_like) = 0;

};

class iDisplayElement
{
public:
	virtual void display() = 0;

};

class iSubject
{
public:
	virtual void registerObserver(iObserver& o) = 0;
	virtual void removeObserver(iObserver& o) = 0;
	virtual void notifyObservers() = 0;
};

class WeatherData : public iSubject {
private:
	std::list<iObserver*> observers;
	float temperature;
	float humidity;

	float min_temperature;
	float max_temperature;

	float pressure;

	float feels_like;

public:
	void registerObserver(iObserver& o)override {
		observers.push_back(&o);

	}

	void removeObserver(iObserver& o) override {
		observers.remove(&o);

	}

	void notifyObservers() override {
		for (auto it = observers.begin(); it != observers.end(); it++) {
			(*it)->Update(temperature, humidity, min_temperature, max_temperature, pressure, feels_like);

		}

	}

	void measurementsChanged() {
		notifyObservers();

	}

	void setMeasurements(float temperature, float humidity, float min_temperature, float max_temperature, float pressure, float feels_like) {
		this->temperature = temperature;
		this->humidity = humidity;
		this->min_temperature = min_temperature;
		this->max_temperature = max_temperature;
		this->pressure = pressure;
		this->feels_like = feels_like;
		measurementsChanged();

	}


};

class CurrentConditionsDisplay : public iObserver, iDisplayElement {
private:
	float temperature;
	float humidity;

	iSubject& weatherData;

public:
	CurrentConditionsDisplay(iSubject& weatherData) : weatherData(weatherData) {
		weatherData.registerObserver(*this);

	}

	void Update(float temperature, float humidity, float min_temperature, float max_temperature, float pressure, float feels_like) override {
		this->temperature = temperature;
		this->humidity = humidity;
		display();

	}

	void display() override {
		std::cout << "Current conditions: " << temperature-273 << "'C degrees and " << humidity << "% humidity" << std::endl;

	}

	void unregister() {
		weatherData.removeObserver(*this);

	}

};

class StatisticDisplay : public iObserver, iDisplayElement {
private:
	float min_temperature;
	float max_temperature;
	iSubject& weatherData;

public:
	StatisticDisplay(iSubject& weatherData) : weatherData(weatherData) {
		weatherData.registerObserver(*this);

	}

	void Update(float temperature, float humidity, float min_temperature, float max_temperature, float pressure, float feels_like) {
		this->min_temperature = min_temperature;
		this->max_temperature = max_temperature;
		display();

	}
	
	void display() override {
		std::cout << "Current min temperature: " << min_temperature - 273 << "'C; average temperature: " << ((min_temperature + max_temperature)/2) - 273
			<< "'C; max temperature: " << max_temperature - 273 << "'C; " << std::endl;

	}

	void unregister() {
		weatherData.removeObserver(*this);
	}

};

class ForecastDisplay : public iObserver, iDisplayElement {
private:
	float pressure;
	iSubject& weatherData;

public:
	ForecastDisplay(iSubject& weatherData) : weatherData(weatherData) {
		weatherData.registerObserver(*this);

	}

	void Update(float temperature, float humidity, float min_temperature, float max_temperature, float pressure, float feels_like) override {
		this->pressure = pressure;
		display();

	}

	void display() override {
		std::cout << "Atmospheric pressure is " << pressure << " millimeters of mercury" << std::endl;

	}

	void unregistred() {
		weatherData.removeObserver(*this);

	}

};

class FeelsLike : public iObserver, iDisplayElement {
private:
	float temperature;
	float feels_like;
	iSubject& weatherData;

public:
	FeelsLike(iSubject& weatherData) : weatherData(weatherData) {
		weatherData.registerObserver(*this);

	}

	void Update(float temperature, float humidity, float min_temperature, float max_temperature, float pressure, float feels_like) override {
		this->temperature = temperature;
		this->feels_like = feels_like;
		display();

	}

	void display() override {
		std::cout << "The actual temperature is " << temperature - 273 << "'C, feels like " << feels_like - 273 << "'C" << std::endl;

	}

	void unregistred() {
		weatherData.removeObserver(*this);

	}
};

int main() {
	WeatherData weatherData;
	CurrentConditionsDisplay currentDisplay(weatherData);
	StatisticDisplay statisticDisplay(weatherData);
	ForecastDisplay forecastDisplay(weatherData);
	FeelsLike feelsLike(weatherData);

	weatherData.setMeasurements(225, 65, 215, 315, 120.0f, 308);
	
	return 0;

}

File: ObserverPython/Source.py
from abc import ABC, abstractmethod
from design import Ui_MainWindow
import requests

class iObserver(ABC):
    @abstractmethod
    def Update(self) : pass

class iSubject(ABC):
    @abstractmethod
    def RegisterObserver(self, observer) : pass
    @abstractmethod
    def RemoveObserver(self, observer) : pass
    @abstractmethod
    def NotifyObservers(self, observer) : pass

class iDisplayElement(ABC):
    @abstractmethod
    def Display(self) : pass

class WeatherData(iSubject):
    def __init__(self):
        self.observers = [] 
        self.temperature = 0
        self.humidity = 0
        self.min_temperature = 0
        self.max_temperature = 0
        self.pressure = 0
        self.feels_like = 0

    def RegisterObserver(self, observer):
        self.observers.append(observer)

    def RemoveObserver(self, observer):
        self.observers.remove(observer)

    def NotifyObservers(self):
        for o in self.observers:
            o.Update(self.temperature, self.humidity, self.min_temperature, self.max_temperature, self.pressure, self.feels_like)

    def measurementsChanged(self):
        self.NotifyObservers()

    def setMeasurements(self, temperature, humidity, min_temperature, max_temperature, pressure, feels_like):
        self.temperature = temperature
        self.humidity = humidity
        self.min_temperature = min_temperature
        self.max_temperature = max_temperature
        self.pressure = pressure
        self.feels_like = feels_like
        self.measurementsChanged()

class CurrentConditionsDisplay(iObserver, iDisplayElement):
    def __init__(self, weatherData, ui):
        self.weatherData = weatherData
        self.ui = ui
        weatherData.RegisterObserver(self)

    def Update(self, temperature, humidity, min_temperature, max_temperature, pressure, feels_like):
        self.temperature = temperature
        self.humidity = humidity
        self.Display()

    def Display(self):
        print ("Current conditions: " + str(self.temperature) + "'C degrees and " + str(self.humidity) + "% humidity ")
        self.ui.set_temp("Temperature: " + str(round(self.temperature - 273, 2)) + "'C or " + str(self.temperature) + "'F ")
        self.ui.set_hum("Humidity: " + str(self.humidity) + "% ")
        

    def Unregister(self):
        self.weatherData.RemoveObserver(self)

class StatisticDisplay(iObserver, iDisplayElement):
    def __init__ (self, weatherData, ui):
        self.weatherData = weatherData
        self.ui = ui
        weatherData.RegisterObserver(self)

    def Update(self, temperature, humidity, min_temperature, max_temperature, pressure, feels_like):
        self.min_temperature = min_temperature
        self.max_temperature = max_temperature
        self.Display()
        

    def Display(self):
        print("Current min temperature: " + str(self.min_temperature) + "'C; average temperature: " + str(((self.min_temperature + self.max_temperature) / 2)) + "'C Max: "  + str(self.max_temperature) +  "'C ")
        self.ui.set_min_temp("Minimal temperature: " + str(round(self.min_temperature - 273, 2)) + "'C or " + str(self.min_temperature) + "'F")
        self.ui.set_ave_temp("Average temperature: " + str(round(((self.min_temperature + self.max_temperature)/2) - 273, 2)) + "'C or " + str(round((self.min_temperature + self.max_temperature)/2, 2)) + "'F")
        self.ui.set_max_temp("Maximal temperature: " + str(round(self.max_temperature - 273, 2)) + "'C or " + str(self.max_temperature) + "'F")

    def Unregister(self):
        self.weatherData.RemoveObserver(self)

class ForecastDisplay(iObserver, iDisplayElement):
    def __init__ (self, weatherData, ui):
        self.weatherData = weatherData
        self.ui = ui
        weatherData.RegisterObserver(self)

    def Update(self, temperature, humidity, min_temperature, max_temperature, pressure, feels_like):
        self.pressare = pressure
        self.Display()
        

    def Display(self):
        print("Atmospheric pressure is " + str(self.pressare) + " millimeters of mercury")
        self.ui.set_pres("Pressure: " + str(self.pressare) + " millimeters of mercury")

    def Unregister(self):
        self.weatherData.RemoveObserver(self)

class FeelsLike(iObserver, iDisplayElement):
    def __init__ (self, weatherData, ui):
        self.weatherData = weatherData
        self.ui = ui
        weatherData.RegisterObserver(self)

    def Update(self, temperature, humidity, min_temperature, max_temperature, pressure, feels_like):
        self.temperature = temperature
        self.feels_like = feels_like
        self.Display()
        

    def Display(self):
        print("The feels like " + str(self.feels_like) + "'C")
        self.ui.set_feels("Feels like: " + str(round(self.feels_like - 273, 2)) + "'C or " + str(self.feels_like) + "'F")

    def Unregister(self):
        self.weatherData.RemoveObserver(self)


